Workflow cache

Egor-S opened the pull request:
Closes #284 

* Add workflow option `cache`
* If cache exists, it would be copied from storage before the run
* Cache is saved after the run to storage
* Cache is unique per-repo & per-user & per-workflow
  * Location `cache/{repo/address}/{user_email}/{workflow_name}/*`
* Add `dstack prune cache` command
  * Add `Backend.delete_workflow_cache`
  * Add `Storage.delete_prefix` to delete trees

Limitations:

* `delete_prefix` is not implemented for hub backend
* `delete_prefix` could be slow for large trees (aws, gcp) — need to use batch operations


Example workflow:

```yaml
workflows:
  - name: cache-demo
    provider: bash
    commands:
      - "du -hd2 /root/.cache"
      - "pip install numpy pandas matplotlib"
      - "du -hd2 /root/.cache"
    cache:
      - "/root/.cache/pip"
```

**Edit:**

Allowed syntax:

```yaml
cache:
  - "/my/cache/foo"
  - path: "my/cache/bar"
```

Egor-S committed:
Add workflow.cache (cli & runner); convert test folders to packages

Egor-S committed:
Download, mount, and upload cache

Egor-S committed:
dstack prune cache: local & remote (no hub)

Egor-S committed:
dstack prune cache: improve parser

Egor-S committed:
runner: handle missing local repo email

Egor-S committed:
CacheSpec: strip dot relative

Egor-S committed:
CacheSpec: fix removepreifx

Egor-S committed:
Leave a note on relative vs absolute mount conflict

Egor-S committed:
cache: update docs

peterschmidt85 reviewed (approved):
Let's add support for Hub and merge. Okay?

peterschmidt85 commented:
For `bash`, `code`, `lab`, and `notebook` providers, let's allow the CLI to automatically replace the `~/` prefix with `/root` (the same place where we remove the `./` prefix). To avoid hardcoding, we can allow the provider to specify whether it is aware of the user's home directory path.
For all others, let's show an error telling that this is not allowed.
This will make defining paths inside the user's home directory much easier.

peterschmidt85 committed:
- [Docs]: Added the `Cache` page (under `Usage`)
- [Docs]: Updated the `Overview`, `Quick start`, and README to use `cache`
- [Docs]: Merged the `Conda` page into `Python` (under `Usage`)
- [Docs]: Minor updates to the YAML reference pages (related to `cache`)

Egor-S committed:
hub: prune workflow cache

Egor-S committed:
upload: add traces

Egor-S committed:
Merge branch 'master' into issue_284_cache

# Conflicts:
#	docs/quick-start.md

Egor-S committed:
upload & download: release MAX_THREADS

Egor-S committed:
Merge branch 'master' into issue_284_cache

Egor-S committed:
cache & artifacts: add home directory expansion with tests

Egor-S commented:
* `hub` backend supports cache pruning
* `aws` backend works correctly with a cache
* `artifacts` and `cache` support `~` (home directory) expansion for dstack-managed providers
